analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs

```diff
@@ -148,15 +161,22 @@ declaration switch
             recursive switch
             {
                 _ when valueConstraint == ObjectConstraint.Null => BoolConstraint.False,
+                _ when RecursivePatternAlwaysMatchesAnyNotNull(recursive) => BoolConstraint.True,
+                _ => null,
+            };
+
+        private static bool RecursivePatternAlwaysMatchesAnyNotNull(IRecursivePatternOperationWrapper recursive) =>
+            recursive switch
+            {
                 {
                     PropertySubpatterns.Length: 0,
                     DeconstructionSubpatterns.Length: 0,
-                } => BoolConstraint.True,
-                _ when SubPatternsAlwaysMatch(recursive) => BoolConstraint.True,
-                _ => null,
+                } => true,
+                _ when recursive.InputType.Equals(recursive.MatchedType) && SubPatternsAlwaysMatches(recursive) => true,
```

Comment:
Use DerivesOrImplements. There might also be the possibility to leverage "NarrowedType == MatchedType" or something like that instead.

Reply:
Add test cases for `someObj is string { }` and `someString is object { }`. Also: with non-empty patterns and some negations.

Reply:
I missed the type check in my PR. Can you add some tests to the "BoolConstraint" tests?